src/software/world/ball_model/linear_ball_model.cpp

```diff
@@ -0,0 +1,113 @@
+#include "software/world/ball_model/linear_ball_model.h"
+
+#include "software/new_geom/util/physics.h"
+
+LinearBallModel::LinearBallModel(BallState initial_ball_state,
+                                 std::optional<FrictionParameters> friction_parameters)
+    : initial_ball_state_(initial_ball_state),
+      friction_parameters_(initFrictionParameters(friction_parameters))
+{
+}
+
+BallState LinearBallModel::estimateFutureState(double seconds_in_future)
+{
+    if (seconds_in_future < 0)
+    {
+        throw std::invalid_argument(
+            "Position estimate is updating to a time in the past");
+    }
+
+    if (friction_parameters_)
+    {
+        return applyLinearFrictionModel(initial_ball_state_, seconds_in_future,
+                                        *friction_parameters_);
+    }
+    else
+    {
+        Point future_position = initial_ball_state_.position() +
+                                initial_ball_state_.velocity() * seconds_in_future;
+        return BallState(future_position, initial_ball_state_.velocity());
+    }
+}
+
+BallState LinearBallModel::applyLinearFrictionModel(
+    const BallState &initial_ball_state, double seconds_in_future,
+    const FrictionParameters &friction_parameters)
+{
+    double initial_velocity_length = initial_ball_state.velocity().length();
```

Comment:
I think we can shrink this function a bit by replacing the branches with `min/max` functions. Here's a (very much untested) snippet that I think could work:
```
    const double initial_speed_m_per_s = initial_ball_state.velocity().length();

    // Figure out how long the ball will roll/slide, if at all
    const double max_sliding_duration_secs =
        (initial_speed_m_per_s - friction_parameters.rolling_sliding_speed_threshold) /
        friction_parameters.sliding_friction_acceleration;
    const double sliding_duration_secs =
        std::max(0, std::min(seconds_in_future, max_sliding_duration_secs));
    const double max_rolling_duration_secs =
        std::min(initial_speed_m_per_s,
                 friction_parameters.rolling_sliding_speed_threshold) /
        friction_parameters.rolling_friction_acceleration;
    const double rolling_duration_secs =
        std::max(0, std::min(seconds_in_future - sliding_duration_secs, max_rolling_duration_secs));

    // Figure out where the ball is after the two stages
    BallState ball_state_after_sliding = calculateFutureBallState(
        initial_ball_state, friction_parameters.sliding_friction_acceleration,
        sliding_duration_secs);
    BallState ball_state_after_rolling = calculateFutureBallState(
        ball_state_after_sliding, friction_parameters.rolling_friction_acceleration,
        rolling_duration_secs);

    return ball_state_after_rolling;
```

Reply:
pretty much copy pasted it